dnshostprovider_test.go

```diff
@@ -172,53 +172,132 @@ func TestDNSHostProviderReconnect(t *testing.T) {
+			{false, true},
+			{false, true},
 
-		// And some more failures.
-		{false, false},
-		{false, false},
-		{true, false}, // Looped back to last known good server: all alternates failed.
-		{false, false},
-	}
+			// And some more failures.
+			{false, false},
+			{false, false},
+			{true, false}, // Looped back to last known good server: all alternates failed.
+			{false, false},
+			{false, false},
+			{true, false},
+			{false, false},
+			{false, false},
+			{true, false},
+			{false, false},
+		}
+
+		for i, td := range testdata {
+			_, retryStartGot := hp.Next()
+			if retryStartGot != td.retryStartWant {
+				t.Errorf("%d: retryStart=%v; want %v", i, retryStartGot, td.retryStartWant)
+			}
+			if td.callConnected {
+				hp.Connected()
+			}
+		}
+	})
 
-	for i, td := range testdata {
-		_, retryStartGot := hp.Next()
-		if retryStartGot != td.retryStartWant {
-			t.Errorf("%d: retryStart=%v; want %v", i, retryStartGot, td.retryStartWant)
+	// The above test for "static" checks that the semantics of the implementation of Next are correct,
+	// this test only checks that the "dynamic" mode correctly interacts with the resolver.
+	t.Run("dynamic", func(t *testing.T) {
+		const fooPort, barPort = "2121", "6464"
+		const fooHost, barHost = "foo.com", "bar.com"
+		hostToPort := map[string]string{
+			fooHost: fooPort,
+			barHost: barPort,
 		}
-		if td.callConnected {
-			hp.Connected()
+		hostToAddrs := map[string][]string{
+			fooHost: {"0.0.0.1", "0.0.0.2", "0.0.0.3"},
+			barHost: {"0.0.0.4", "0.0.0.5", "0.0.0.6"},
 		}
-	}
+		addrToHost := map[string]string{}
+		for host, addrs := range hostToAddrs {
+			for _, addr := range addrs {
+				addrToHost[addr+":"+hostToPort[host]] = host
+			}
+		}
+
+		hp := &DNSHostProvider{
+			static: false,
+			lookupHost: func(host string) ([]string, error) {
+				addrs, ok := hostToAddrs[host]
+				if !ok {
+					t.Fatalf("Unexpected argument to lookupHost %q", host)
+				}
+				return addrs, nil
+			},
+		}
+
+		err := hp.Init([]string{fooHost + ":" + fooPort, barHost + ":" + barPort})
+		if err != nil {
+			t.Fatalf("Unexpected err from Init %v", err)
+		}
+
+		addr1, retryStart := hp.Next()
+		if retryStart {
+			t.Fatalf("retryStart should be false")
+		}
+		addr2, retryStart := hp.Next()
+		if !retryStart {
+			t.Fatalf("retryStart should be true")
+		}
+		host1, host2 := addrToHost[addr1], addrToHost[addr2]
+		if host1 == host2 {
+			t.Fatalf("Next yielded addresses from same host (%q)", host1)
+		}
+
+		// Final sanity check that it is shuffling the addresses
+		seenAddresses := map[string]map[string]bool{
+			fooHost: {},
+			barHost: {},
+		}
+		for i := 0; i < 10_000; i++ {
+			addr, _ := hp.Next()
+			seenAddresses[addrToHost[addr]][addr] = true
+		}
+
+		for host, addrs := range hostToAddrs {
+			for _, addr := range addrs {
+				if !seenAddresses[host][addr+":"+hostToPort[host]] {
+					t.Fatalf("expected addr %q for host %q not seen (seen: %v)", host, addr, seenAddresses)
```

Comment:
host and addr params should be reversed.

Reply:
Nice catch!